File: PHP/0046_permutations.php
<?php
/*=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
 * 46. Permutations
 * 
 * Medium
 * 
 * Given a collection of distinct integers, return all possible permutations.
 * 
 * Example:
 * Input: [1,2,3]
 * Output:
 * [
 *   [1,2,3],
 *   [1,3,2],
 *   [2,1,3],
 *   [2,3,1],
 *   [3,1,2],
 *   [3,2,1]
 * ]
 ** =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=*/

class Solution {

    /**
     * @param Integer[] $nums
     * @return Integer[][]
     */
    function permute($nums) {
        $res = [];
        if ($nums == null || count($nums) == 0) return $res;
        $currentX = [];
        $used = array_fill(0, count($nums), false);
        for($j=0; $j<count($used);$j++)
        {
            $used[$j] = false;
        }
        self::bt($currentX, $nums, $used, $res);
        return $res;
    }

        function bt($currentX, &$nums, $used, &$res)
        {
            if (count($currentX) == count($nums))
            {
                array_push($res, $currentX);
                return;
            }

            for ($i = 0; $i < count($nums); $i++)
            {

                if ($used[$i] == true) continue;
                array_push($currentX, $nums[$i]);
                $used[$i] = true;

                self::bt($currentX, $nums, $used, $res);

                array_pop($currentX);
                $used[$i] = false;
            }
        }
}

File: Python/0046_permutations.py
"""=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
 * 46. Permutations
 * 
 * Medium
 * 
 * Given a collection of distinct integers, return all possible permutations.
 * 
 * Example:
 * Input: [1,2,3]
 * Output:
 * [
 *   [1,2,3],
 *   [1,3,2],
 *   [2,1,3],
 *   [2,3,1],
 *   [3,1,2],
 *   [3,2,1]
 * ]
 ** =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-="""

 class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        
        if len(nums) == 0:
        	return []
        if len(nums) == 1:
        	return [nums]

        result = []
        for index in range(len(nums)):
        	for p in self.permute(nums[0:index] + nums[index+1:]):
        		result.append([nums[index]] + p)

        return result